### PR TITLE
Progress bar improvements

### PR DESCRIPTION
**Pull request by otaviobp**

### CODE REVIEW

File: src/update.c [hunk 1]
```diff
@@ -105,10 +105,8 @@ static int check_manifests_uniqueness(int clrver, int mixver)
 static int update_loop(struct list *updates, struct manifest *server_manifest)
 {
 	int ret;
-	struct step step;
 
-	step = progress_get_step();
-	progress_set_step(step.current, "download_fullfiles");
+	progress_next_step("download_fullfiles", PROGRESS_BAR);
```
Comment (jwakre) [inline src/update.c]: @otaviobp  When the statedir is populated (previously existing files, pack extraction) this step reports 0 progress for ~20 seconds followed by an immediate jump to 100. This time is spent validating the hashes of fullfiles in the statedir and cache hits (no download) do not impact the progress bar. I feel this is too long to sit without feedback.

To solve this problem, we could improve the `check_files_hash` step to validate files in the statedir in addition to the files it checks at the target install locations. This step could create a list of files that we need to download (not already installed and not present in statedir) which could be passed to `download_fullfiles`. In this case, all files passed to `download_fullfiles` would require downloads.

Comment (otaviobp) [reply src/update.c]: Easier for now would be to create another step inside download_fullfiles to process that. Do you think it will be a good solution?

Comment (jwakre) [reply src/update.c]: It would work, but it seems weird to have back-to-back file hash verifying steps.

Comment (jwakre) [reply src/update.c]: But I can accept that

Comment (jwakre) [reply src/update.c]: This is working well for the installer after adding a step.

File: src/staging.c [hunk 2]
```diff
@@ -397,6 +400,7 @@ which ends after rename_all_files_to_final() succeeds
 			return ret;
 		}
 	}
+	progress_report(complete, list_length * 2);
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/staging.c]: staging_install_all_files() runs 2 steps, staging and moving. We were reporting progress for moving, but not staging. In the average use case staging is slower than moving so we are reporting progress for both at the same step. We could add another step for this, but I think it's better to have just one for now.

File: src/bundle.c [hunk 3]
```diff
@@ -609,8 +609,9 @@ enum swupd_code remove_bundles(struct list *bundles)
 
 		if (list_len(files_to_remove) > 0) {
 			info("\nDeleting bundle files...\n");
-			progress_set_next_step("remove_files");
+			progress_next_step("remove_files", PROGRESS_BAR);
 			int deleted = remove_files_from_fs(files_to_remove);
+			progress_next_step("complete_remove", PROGRESS_UNDEFINED);
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle.c]: We don't need this. I'll remove

File: src/bundle_add.c [hunk 4]
```diff
@@ -529,20 +526,9 @@ enum swupd_code main_bundle_add()
 enum swupd_code bundle_add_main(int argc, char **argv)
 {
 	int ret;
+	/* 5 steps + 2 extras for pack and fullfile */;
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle_add.c]: There are steps inside pack and fullfile code.

File: src/bundle_remove.c [hunk 5]
```diff
@@ -98,7 +98,7 @@ enum swupd_code bundle_remove_main(int argc, char **argv)
 {
 	struct list *bundles_list = NULL;
 	int ret;
-	const int steps_in_bundleremove = 1;
+	const int steps_in_bundleremove = 3;
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle_remove.c]: same as above.

File: src/bundle_add.c [hunk 6]
```diff
@@ -322,25 +320,21 @@ static enum swupd_code install_bundles(struct list *bundles, struct manifest *mo
...
-	progress_complete_step();
 	timelist_timer_stop(globals.global_times); // closing: Consolidate files from bundles
 
 	if (!to_install_files) {
 		goto out;
 	}
 
-	///* Step 3: Check if we have enough space */
-	progress_set_step(3, "check_disk_space_availability");
+	///* Check if we have enough space */
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle_add.c]: Fixed.

File: src/bundle_add.c [hunk 7]
```diff
@@ -533,19 +524,7 @@ enum swupd_code main_bundle_add()
...
-	 *
-	 *  1) load_manifests
-	 *  2) consolidate_files
-	 *  3) check_disk_space_availability
-	 *  4) download_packs
-	 *  5) download_fullfiles
-	 *  6) install_files
-	 *  7) progress_set_step
-	 */
+	const int steps_in_bundleadd = 8;
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle_add.c]: Maintaining the number of steps is hard anyway. My suggestion is actually to remove the total number of steps because the installer isn't using it. I'm just waiting this patch to get in to suggest the other change.

File: src/verify.c [hunk 8]
```diff
@@ -807,46 +805,9 @@ enum swupd_code verify_main(void)
...
-	if (cmdline_option_picky || cmdline_option_extra_files_only) {
-		steps_in_verify += 1;
-	}
-	if (cmdline_option_quick) {
-		steps_in_verify -= (cmdline_option_picky || cmdline_option_extra_files_only) ? 3 : 2;
-	}
-	if (cmdline_option_extra_files_only) {
-		steps_in_verify = 5;
-	}
-	progress_init_steps("verify", steps_in_verify);
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/verify.c]: Instead of setting that for verify I'm now setting it for each individual user of verify (os-install, diagnose, repair)

File: src/fullfile.c [hunk 9]
```diff
@@ -189,11 +195,13 @@ int download_fullfiles(struct list *files, int *num_downloads)
 
 	if (!need_download) {
 		/* no file needs to be downloaded */
+		progress_next_step("download_fullfiles", PROGRESS_BAR);
```
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/fullfile.c]: Good point. I'll fix it

**Comment by karthikprabhuvinod** [comment]
`sudo ./swupd os-install -B os-core,os-core-update --path=/home/user/test_os_install
Warning: /home/user/test_os_install//usr/share/clear/versionstamp does not exist!
Installing OS version 31300 (latest)

Downloading missing manifests...

Downloading packs (35.53 MB) for:
 - os-core-plus
 - libstdcpp
 - libglib
 - findutils
 - p11-kit
 - os-core-webproxy
 - os-core
 - os-core-update
 [100%]

Finishing packs extraction...

Checking for corrupt files
 [100%]


 [100%]

No extra files need to be downloaded

Installing base OS and selected bundles
 [100%]


Inspected 5136 files
  5026 files were missing
    5026 of 5026 missing files were installed
    0 of 5026 missing files were not installed

Calling post-update helper scripts
Warning: helper script (/home/user/test_os_install//usr/bin/clr-boot-manager) not found, it will be skipped

Installation successful`

**Comment by karthikprabhuvinod** [comment]
@otaviobp Checking for corrupt files gave two 100%. We need to print something before that.

**Comment by otaviobp** [comment]
Yes, you're right, I'll fix that

File: src/bundle_add.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (castulo) [reply src/bundle_add.c]: the initial idea of having the total number of steps was because that is the only way to be able to show real progress, otherwise you can only show the progress of the current step but it is impossible to know the progress of the whole process. 

So my idea originally was that the installer (or perhaps a GUI tool for swupd in the future) was going to show an overall progress of the whole swupd operation, while the percentages in a step could be used to show progress of the current action. That way you could have something like:
```
$swupd bundle-add BUNDLE

|------>           |   Overall bundle-add progress
|----------------->|     - loading manifests
|----------------->|     - downloading packs
|----------->      |     - installing files
```

At the moment the number of steps have not been used by any gui, but I don't think we should get rid of it in case at some point some tool wants to show progress like that.

File: src/verify.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (castulo) [reply src/verify.c]: got it

File: src/bundle_add.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/bundle_add.c]: ok. I won't remove that then. And I'll add comments for all the steps back.

**Comment by otaviobp** [comment]
Fixed everything pending. @castulo @karthikprabhuvinod Can you take a look?

**Comment by karthikprabhuvinod** [comment]
I wanted to try a big package so I ran the below

sudo ./swupd os-install -B os-core-update,big-data-basic --path=<PATH>/test_os_it --json-output
[
{ "type" : "start", "section" : "os-install" },
{ "type" : "warning", "msg" : "/home/kp/swupd-client/test_os_it//usr/share/clear/versionstamp does not exist!" },
{ "type" : "progress", "currentStep" : 1, "totalSteps" : 9, "stepCompletion" : -1, "stepDescription" : "load_manifests" },
{ "type" : "info", "msg" : "Installing OS version 31310 (latest)" },
{ "type" : "info", "msg" : "Downloading missing manifests..." },
{ "type" : "progress", "currentStep" : 1, "totalSteps" : 9, "stepCompletion" : 100, "stepDescription" : "load_manifests" },
**{ "type" : "progress", "currentStep" : 2, "totalSteps" : 9, "stepCompletion" : 0, "stepDescription" : "download_packs" },**
{ "type" : "info", "msg" : "Downloading packs (32.46 MB) for:" },
{ "type" : "info", "msg" : " - elasticsearch" },
**{ "type" : "progress", "currentStep" : 2, "totalSteps" : 9, "stepCompletion" : 0, "stepDescription" : "download_packs" },**

I saw a the same progress being repeated twice..I am guessing this should happen

File: src/verify.c [hunk 10]
```diff
@@ -250,19 +250,16 @@ static int get_required_files(struct manifest *official_manifest, struct list *s
 	int ret;
 
 	if (cmdline_option_install) {
-		progress_set_next_step("download_packs");
 		print("\n");
 		get_all_files(official_manifest, subs);
 	}
 
-	progress_set_next_step("check_files_hash");
+	progress_next_step("check_files_hash", PROGRESS_BAR);
```
Comment (karthikprabhuvinod) [inline src/verify.c]: Just a thought:

Isnt this approach going to add more variability as to the dev deciding whether its going to  PROGRESS_BAR or PROGRESS_UNDEFINED. 

Maybe its too late and a little expensive to think of this, I was thinking maybe we should
use the timer library of c which calls our callback at regular intervals of time.

What I am hinting at is starting a timer at start of an operation and let the kernel calls our callback @ regular intervals for printing progress and then close the progress.

Comment (otaviobp) [reply src/verify.c]: > Isnt this approach going to add more variability as to the dev deciding whether its going to PROGRESS_BAR or PROGRESS_UNDEFINED.

I didn't understand your comment. Can you rephrase?

> Maybe its too late and a little expensive to think of this, I was thinking maybe we should
use the timer library of c which calls our callback at regular intervals of time.

It's not too late and it didn't block this PR. I'm ok on improving the spinner to be displayed even when there's not internet download. It's better than what we have now. But it's low priority though.

Comment (karthikprabhuvinod) [reply src/verify.c]: What I meant there can be a confusion about an operation whether it is need to report PROGRESS_BAR or PROGRESS_UNDEFINED.  But don't feel strongly about this as at-least now there is a way to classify an operation.

**Comment by otaviobp** [comment]
Fixed the @karthikprabhuvinod reported problems

**Comment by karthikprabhuvinod** [comment]
@otaviobp I tested on-install & verify and it looks good to me. Great stuff.

**Comment by mdhorn** [comment]
@otaviobp @jwakre 
A new release of clr-installer was made today which supports these changes
https://github.com/clearlinux/clr-installer/releases/tag/2.3.2
It was pushed to koji and will be in tonight's Clear Linux OS build.
We should be good to have a new swupd release.
